File: Adapters/ResearchFeatureEngine.MT5/RFEMT5Bridge.cpp
//+------------------------------------------------------------------+
//| RFEMT5Bridge.cpp — Native C ABI bridge for MQL5 → .NET Core engine |
//+------------------------------------------------------------------+
// This file is NOT compiled directly by MSBuild. It is a template for
// the native bridge DLL that MQL5 #imports. The actual build is done by
// a separate C++/CLI or C# UnmanagedCallersOnly project that exports
// C ABI functions callable from MQL5 via #import.
//
// MQL5 CANNOT import .NET assemblies directly via #import. The bridge
// must be a native Win32 DLL that internally hosts the .NET runtime
// (via nethost + coreclr hosting API) or uses C++/CLI on Windows.
//
// Build (when MT5 terminal is installed):
//   cl /LD /EHsc RFEMT5Bridge.cpp /link /OUT:RFEMT5Bridge.dll
//
// Or via C# with UnmanagedCallersOnly (net8.0+):
//   dotnet publish -r win-x64 -p:NativeLib=True
//
// Exported functions (C ABI, __stdcall calling convention):
//
// int RfeInitialize(
//     const char* symbol,         // _Symbol as ANSI string
//     int timeframe,              // (int)_Period (ENUM_TIMEFRAMES)
//     int ref_type,               // ENUM_REFERENCE_TYPE
//     int atr_period,
//     double atr_multiplier,
//     int smooth_length,
//     int box_length,
//     int hma_period,
//     int scale_atr_period,
//     int stats_window,
//     int reversal_mode
// );
//
//   Returns: 0 on success, non-zero error code on failure.
//
// bool RfeProcessBar(
//     double open, double high, double low,
//     double close, long volume, long time
// );
//
//   Returns: true if the engine produced valid output for this bar,
//   false if still in warm-up.
//
// double RfeGetReference();  // EngineValues.Reference.Price
// double RfeGetDistance();   // DirectionalDistance (close − reference)
// double RfeGetScale();      // Scale (ATR)
// double RfeGetNormalized();  // Normalization (distance / scale)
//
// void RfeShutdown();
//
// Architecture:
//   MQL5 → #import RFEMT5Bridge.dll (native C ABI)
//        → RfeInitialize creates a .NET Core host (via nethost)
//        → Loads ResearchFeatureEngine.dll + ResearchFeatureEngine.Python.dll
//        → Creates PythonEngineFactory.RunAll or equivalent
//        → Each RfeProcessBar feeds one bar through engine.Update()
//        → RfeGet* retrieve values from EngineValues
//
// No engine math is duplicated — all values originate from the canonical
// EngineValues pipeline in the Core assembly.
//
// Dependencies:
//   - nethost.dll (ships with .NET 6+ runtime)
//   - hostfxr.dll (ships with .NET 6+ runtime)
//   - ResearchFeatureEngine.dll (Core)
//   - ResearchFeatureEngine.Python.dll (adapter for type resolution)
//       (NOT ResearchFeatureEngine.MT5.dll — the bridge calls Core directly)
//
// The MT5 adapter project (Adapters/ResearchFeatureEngine.MT5/) compiles
// MT5MarketData.cs as a .NET assembly that is used by the bridge via
// reflection or direct reference. The bridge itself is a separate native
// project not included in this repository (requires C++/CLI toolchain).

File: Adapters/ResearchFeatureEngine.MT5/RFEMT5Bridge/RFEMT5Bridge.cs
//+------------------------------------------------------------------+
//| RFEMT5Bridge.cs — Native C ABI bridge for MQL5 -> .NET Core       |
//+------------------------------------------------------------------+
// This is a STUB implementation for COMPILE-VALIDATION purposes.
// It exports the same C ABI functions that RFEMT5Bridge.cpp documents,
// but returns NaN/0 sentinel values. The real bridge (documented in
// RFEMT5Bridge.cpp) would host the .NET Core engine via nethost and
// delegate to the canonical EngineValues pipeline.
//
// Build: dotnet publish -c Release -r win-x64 --self-contained false -p:NativeLib=true
//   Output: bin/Release/net8.0/win-x64/publish/RFEMT5Bridge.dll
//
// This stub allows the MQL5 indicator (RFEMT5Indicator.mq5) to compile
// in MetaEditor. At runtime, replace this stub with the full bridge
// documented in RFEMT5Bridge.cpp.
//
// Default calling convention is StdCall on Windows (MQL5 #import expects
// __stdcall, which is the default for UnmanagedCallersOnly in .NET 8
// Native AOT on Windows).

using System;
using System.Runtime.InteropServices;

namespace RFEMT5Bridge
{
    internal static class NativeExports
    {
        // EngineValues for the current bar
        private static double s_reference = double.NaN;
        private static double s_distance = double.NaN;
        private static double s_scale = double.NaN;
        private static double s_normalized = double.NaN;
        private static int s_initialised = 0;

        /// <summary>
        /// Initializes the ResearchFeatureEngine MT5 bridge.
        /// </summary>
        /// <returns>0 on success, non-zero error code on failure.</returns>
        [UnmanagedCallersOnly]
        public static int RfeInitialize(
            IntPtr symbol,           // ANSI string (_Symbol)
            int timeframe,           // (int)_Period
            int ref_type,            // ENUM_REFERENCE_TYPE
            int atr_period,
            double atr_multiplier,
            int smooth_length,
            int box_length,
            int hma_period,
            int scale_atr_period,
            int stats_window,
            int reversal_mode)
        {
            // Stub: always succeed
            s_initialised = 1;
            s_reference = double.NaN;
            s_distance = double.NaN;
            s_scale = double.NaN;
            s_normalized = double.NaN;
            return 0;
        }

        /// <summary>
        /// Processes one bar through the engine pipeline.
        /// Returns true if valid output was produced, false during warm-up.
        /// </summary>
        [UnmanagedCallersOnly]
        [return: MarshalAs(UnmanagedType.I1)]
        public static bool RfeProcessBar(
            double open,
            double high,
            double low,
            double close,
            long volume,
            long time)
        {
            // Stub: always return false (warm-up) so indicator doesn't plot garbage
            return false;
        }

        [UnmanagedCallersOnly]
        public static double RfeGetReference() => s_reference;

        [UnmanagedCallersOnly]
        public static double RfeGetDistance() => s_distance;

        [UnmanagedCallersOnly]
        public static double RfeGetScale() => s_scale;

        [UnmanagedCallersOnly]
        public static double RfeGetNormalized() => s_normalized;

        [UnmanagedCallersOnly]
        public static void RfeShutdown()
        {
            s_initialised = 0;
        }
    }
}
